server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java

```diff
@@ -24,73 +24,171 @@
                                           .build();
     }
 
-    private String pattern(String type) {
+    private String pattern(String type, String[] esmessagefields) {
         if (Strings.isEmpty(type)) {
             throw new IllegalArgumentException("layout parameter 'type_name' cannot be empty");
         }
-        return PATTERN.replace("${TYPE}", type);
+        Map<String, Object> map = new LinkedHashMap<>();
+        map.put("type", inQuotes(type));
+        map.put("timestamp", inQuotes("%d{yyyy-MM-dd'T'HH:mm:ss,SSSZZ}"));
+        map.put("level", inQuotes("%p"));
+        map.put("component", inQuotes("%c{1.}"));
+        map.put("cluster.name", inQuotes("${sys:es.logs.cluster_name}"));
+        map.put("node.name", inQuotes("%node_name"));
+        map.put("message", inQuotes("%notEmpty{%enc{%marker}{JSON} }%enc{%.-10000m}{JSON}"));
+
+        for (String key : esmessagefields) {
+            map.put(key, "%ESMessageField{" + key + "}");
+        }
+
+        return createPattern(map);
+    }
+
+    private String createPattern(Map<String, Object> map) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("{");
+        String separator = "";
+        for (Map.Entry<String, Object> entry : map.entrySet()) {
+            sb.append(separator);
+            sb.append(jsonKey(entry.getKey()));
+            sb.append(entry.getValue().toString());
+            separator = ", ";
+        }
+        sb.append(notEmpty(", %node_and_cluster_id "));
+        sb.append("%exceptionAsJson ");
+        sb.append("}");
+        sb.append(System.lineSeparator());
+
+        return sb.toString();
+    }
+
+    private String notEmpty(String value) {
+        return "%notEmpty{" + value + "}";
+    }
+
+    private CharSequence jsonKey(String s) {
+        return inQuotes(s) + ": ";
+    }
+
+    private String inQuotes(String s) {
+        return "\"" + s + "\"";
     }
 
     @PluginFactory
-    public static ESJsonLayout createLayout(@PluginAttribute("type_name") String type,
-                                            @PluginAttribute(value = "charset", defaultString = "UTF-8") Charset charset) {
-        return new ESJsonLayout(type, charset);
+    public static ESJsonLayout createLayout(String type,
+                                            Charset charset,
+                                            String[] esmessagefields) {
+        return new ESJsonLayout(type, charset, esmessagefields);
+    }
+
+    public static class Builder<B extends ESJsonLayout.Builder<B>> extends AbstractStringLayout.Builder<B>
+        implements org.apache.logging.log4j.core.util.Builder<ESJsonLayout> {
+
+        @PluginAttribute("type_name")
+        String type;
+
+        @PluginAttribute(value = "charset", defaultString = "UTF-8")
+        Charset charset;
+
+        @PluginAttribute("esmessagefields")
+        private String esMessageFields;
+
+        public Builder() {
+            super();
```

Comment:
explicit call to super should not be necessary

Reply:
Looks like this wasn't addressed

Reply:
somehow I missed that - fixed now